src/retrieve/uuid.rs

```diff
@@ -0,0 +1,53 @@
+use cis_client::getby::GetBy;
+use cis_client::AsyncCisClientTrait;
+use failure::Error;
+use futures::future::Either;
+use futures::future::IntoFuture;
+use futures::Future;
+use log::info;
+use log::warn;
+use lru_time_cache::LruCache;
+use std::sync::Arc;
+use std::sync::RwLock;
+
+pub fn get_uuid<T: AsyncCisClientTrait>(
+    user_id: &str,
+    cis_client: &T,
+    cache: &Arc<RwLock<LruCache<String, String>>>,
+    own: bool,
+) -> impl Future<Item = Option<String>, Error = Error> {
+    if !own {
+        return Either::A(Ok(None).into_future());
+    }
+    let user_id_f = user_id.to_owned();
+    let cache_f = Arc::clone(cache);
+
+    if let Some(Some(uuid)) = cache
+        .write()
+        .ok()
+        .map(|mut c| c.get(user_id).map(Clone::clone))
+    {
+        Either::A(Ok(Some(uuid)).into_future())
+    } else {
+        Either::B(
+            cis_client
+                .get_user_by(user_id, &GetBy::UserId, None)
+                .map_err(Into::into)
+                .map(|p| p.uuid.value)
+                .map(move |uuid| {
+                    if let Some(uuid) = uuid {
+                        let msg = format!("updated cache for {}", &user_id_f);
+                        cache_f
+                            .write()
+                            .ok()
+                            .map(|mut c| c.insert(user_id_f, uuid.clone()));
```

Comment:
Is it worth storing this `user_id` → `uuid` map in a more permanent store?

How expensive is a call to CIS for this? Grepping through the code it looks like our cache is for one day, so maybe that's not too bad for the first page load each day (for a small subset of users).

I assume you also didn't use a more permanent store because Auth0 account linking can change the `user_id` → `uuid` mapping and caching means that's a problem for up to a day rather than all eternity.

Reply:
Yeah that's keeps to be an upcoming issue. I propose we switch to a proxy which adds this to the header (I started sketching this).